### ISSUE TITLE
specialComments and @media queries

### ISSUE THREAD

**Issue by jamesmacwhite** [jakubpawlowicz/clean-css#890]
I've been testing clean-css 4 in a project that I use it in.

I've noticed when using `specialComments` `0` inside code like this:

```
@media screen { 
/*!
    A special CSS comment
*/
}
```

The whole media query block is removed. Previously under clean-css 3, Having the same special comment within a media query block without any actual CSS rules (essentially an empty media query) would prevent this from happening, and the following would be left over.

```
@media screen {
}
```

I'm aware of some of the configuration changes between the versions, is there a specific option that's causing this, that I can control?

**Comment by jakubpawlowicz** [comment]
That's because in 3.x all special content, like comments or quoted text, was escaped before optimizations and restored / removed if needed after all optimizations, so your `@media` block was effectively:

```css
@media screen{__ESCAPED_COMMENT_CLEAN_CSS(...)__}
```

during optimizations which prevented it from being removed.

In 4.x all escaping is gone and comments are removed during optimizations which also makes removing then empty `@media` blocks possible.

There is no switch to control it, except `specialComments` flag, as this behavior is consistent with what browsers do.

**Comment by jamesmacwhite** [comment]
@jakubpawlowicz 

Thanks for your reply and explanation.

This is how I am setting up media queries to be included in a email boilerplate template:

https://github.com/centralcollegenottingham/HTML-Email-Boilerplate-Redux/blob/master/app/css/media-queries.css

I used the special comments as essentially "data" within the media query so it didn't get stripped, otherwise exactly the same behaviour I have now under clean-css 4 would happen under version 3.

However based on what you've said I'm going to have to rethink this approach, unless its possible to define it as a enhancement/feature to allow empty media queries.

**Comment by jakubpawlowicz** [comment]
So as far as I understand you optimize `media-queries.css` and then add styling inside `@media`, so, because clean-css 4 removes those rules, you can't do it anymore, right?

Have a look at level 2 optimizations - https://github.com/jakubpawlowicz/clean-css/blob/master/lib/optimizer/level-2/optimize.js#L67 - `removeEmpty` is the only one without an on/off switch, as I've never considered the switch to be useful.

**Comment by jamesmacwhite** [comment]
Right. I have various media queries stored in the that, that will get merged into the final HTML document. This works for the guidelines version of the boilerplate, but not the commentless version now.

I have both level 1 and 2 flags set to false, but sadly it doesn't work anymore, its all down to specialComments being 0.

**Comment by jakubpawlowicz** [comment]
I think we could have a switch for flipping `removeEmpty` on and off.

What bothers me is where to have it, as both level 1 and 2 has code removing empty elements (since any of them can be turned off), so the flip needs to be on both levels too, e.g.

```js
new CleanCSS({ level: { 1: { removeEmpty: false } } }).minify(...)
```

will use level 1 optimizations and not remove empty rules, while

```js
new CleanCSS({ level: { 1: { removeEmpty: false }, 2: { removeEmpty: false } }).minify(...)
```

will skip removing empty rules on both levels.

Otherwise we'd need to introduce a new option and remove empty after level 1 and 2 optimizations, e.g.

```js
new CleanCSS({ removeEmpty: false, level: 2 }).minify(...)
```

which is going back to top level options we had in 3.x which I'd like to avoid.

**Comment by jamesmacwhite** [comment]
That would be ideal, I just need something to be able to allow for empty media queries to be kept in the output under clean-css 4.

**Comment by jakubpawlowicz** [comment]
The thing is it'd apply to all elements not just media queries.

**Comment by jamesmacwhite** [comment]
In my case, I think that would be OK, as that does make sense. Media queries are just the main example in my case.

**Comment by jakubpawlowicz** [comment]
^ there's the implementation - pending mental review :-)

**Comment by jamesmacwhite** [comment]
@jakubpawlowicz Many thanks! Look forward to testing it and implementing it into my project!

**Comment by jakubpawlowicz** [comment]
I've pushed it to master. In the meantime before 4.1 is released you are more than welcome to use the master branch or pin your dependencies to the a5a666f commit.

**Comment by jamesmacwhite** [comment]
@jakubpawlowicz Thanks! I actually use `gulp-clean-css` which acts as a wrapper, but I can manually test it by overriding the clean-css dep to a specific commit/branch.

**Comment by jakubpawlowicz** [comment]
It's been released in 4.1 as described above.
